### ISSUE TITLE
Implement *-reverse flex-directions

### ISSUE THREAD

**Issue by danlrobertson** [servo/servo#10181]
Currently [`flex-direction`](https://drafts.csswg.org/css-flexbox/#flex-direction-property) values of `column-reverse` and `row-reverse` are treated as their non-reversed counterparts.

**Comment by shinglyu** [comment]
I want to give it a try.

**Comment by danlrobertson** [comment]
Awesome! Go for it! There is a lot of work to be done on flex, so any help is appreciated. The spec is https://drafts.csswg.org/css-flexbox.

A previous attempt to implement this merely reversed the `items` vector, but after some thought I don't think that is the correct approach. Since posting this I've realized there are quite a few steps that might need to be implemented before we can correctly implement `*-reverse`. For instance, [`line-sizing`](https://drafts.csswg.org/css-flexbox/#line-sizing). Thoughts or comments on the prior two points? This also may be blocked by #10178?

**Comment by shinglyu** [comment]
@danlrobertson Thanks for the information. Can you elaborate on why reversing the vector is not correct? Any particular case that will violate the spec?

I was thinking about reversing the direction of offset calculation. (I haven't read the code yet. Just some intuition.) Something like 
```
left-offset = offset
```
to
```
left-offset = container.width - ( offset + self.width)
```
Does that make sense at all?

**Comment by danlrobertson** [comment]
@shinglyu: no problem. The section in the spec on [`flex-direction`](https://drafts.csswg.org/css-flexbox/#propdef-flex-direction) states the following for `row-reverse` and `column-reverse`

> Same as (row|column), except the main-start and main-end directions are swapped. 

If you just reverse the order of the items to be placed the location of `main-start` doesn't change, so you're items are still aligned on the same side as if the `flex-direction` were not reversed. The ordering has changed but the alignment hasn't.

Checkout [this example](http://danrobertson.org/examples/flex-direction) I just quickly created to see the effect of the various flex directions.

So your pseudo-code seems more correct.

**Comment by shinglyu** [comment]
I see. I'll start doing some experiment on it. I'm not familiar with this part of Servo, so it might take some time. Hope it won't block any other critical stuff.

**Comment by shinglyu** [comment]
I got some assignment on Servo infra stuff. May not have the bandwidth to work on this right now.

**Comment by shinglyu** [comment]
@danlrobertson I'm trying to trace the code for flexbox, but I only got so far:
```
?
=> construct.rs:PostorderNodeMutTraversal for FlowConstructor 
  => (construct.rs) FlowConstructor::build_flow_for_flex()
    => flow = (flex.rs) FlexFlow::from_fragment()
  => FlowConstructor::build_flow_for_block_like(flow)    
    => build_flow_for_block_starting_with_fragments(flow)
      => ...
```
Not quite sure how it was invoked, and how the functions in `FlexFlow` are triggered. I might need a walkthrough for the FlexFlow's construction.

I found that in `construct.rs`, flex-related called are often delagated to functions for `_block_like`, but in `FlexFlow` there is this logic:
```
110     ¦   let main_mode = match fragment.style.get_position().flex_direction {     
111     ¦   ¦   flex_direction::T::row_reverse | flex_direction::T::row => Mode::Inline,
112     ¦   ¦   flex_direction::T::column_reverse | flex_direction::T::column => Mode::Block
113     ¦   }; 
``` 
which make me very confused. 

Anyway, I guess the code I need to modify is something similar to https://mxr.mozilla.org/servo/source/components/layout/flex.rs#419 , am I right?

**Comment by danlrobertson** [comment]
No problem! I'm still a bit of a newcomer, so take everything I say with a grain of salt.

You're correct that the `FlexFlow` is constructed in `construct.rs`, but the main methods for `Flow` that compute the `inline` and `block` size are in [`flow.rs:183-204`](https://github.com/servo/servo/blob/master/components/layout/flow.rs#L183-L204), which are `impl`ed for `FlexFlow` as you have noted above.

The `main_mode` indicates which axis (inline or block) has the same orientation as the [main axis](https://drafts.csswg.org/css-flexbox/#main-axis). You will notice that a good chunk of the work for each of the passes for `FlexFlow` is actually accomplished in `FlexFlow::(inline|block)_mode_<pass-name-here>`. So if the `main_mode` is `Inline` then most of the work to assign inline sizes will actually be done in `inline_mode_assign_inline_sizes`. Take a look at the last lines of `FlexFlow`s impl of each pass for more details. This works well because we currently only support two directions `row` and `column`. I am still debating as to whether or not this is the best way forward.

So to the best of my knowledge, the code that needs to be modified should be in `inline_mode_assign_inline_sizes` for `row_reverse` and `block_mode_assign_block_size` for `column_reverse`.

Hint: Take a look at `inline_child_start` in `inline_mode_assign_inline_sizes` and `cur_b` in `block_mode_assign_block_size`.

**Comment by shinglyu** [comment]
@danlrobertson : Thank you, that was a very helpful explanation.

**Comment by shinglyu** [comment]
@danlrobertson What flexbox property does servo support right now?
I tried to open this [test page](https://gist.github.com/shinglyu/0e3bc00ca2835354c8a27c92007ff281), but all the 4 flex-direction gives me this kind of layout: 
![_087](https://cloud.githubusercontent.com/assets/3250983/14845520/51f2e9ce-0c91-11e6-9c14-7a96f2cd0821.png)

I looked into the Rust log and found

```
INFO:script::reporter: 4:5 Unsupported property declaration: 'display: flex;'
INFO:script::reporter: 1:1 Unsupported property declaration: 'flex-direction: row;'
```

**Comment by danlrobertson** [comment]
Ah! Good question! A lot of flex style properties are experimental at this time, so they have to be enabled. Use something like the following

```
   ./mach rud -d -- --pref layout.flex.enabled --pref layout.flex-direction.enabled ...
```

**Comment by jdm** [comment]
See the `layout` preferences starting at http://mxr.mozilla.org/servo/source/resources/prefs.json#43 .

**Comment by shinglyu** [comment]
Wow, I didn't know that. Is that documented anywhere?

**Comment by jdm** [comment]
I don't believe so. If you can come up with a useful place to document it, I'm open to the idea :)

**Comment by shinglyu** [comment]
Is wiki a suitable place? The pages don't look very structured, so I wonder if we want to make it a go-to place for documentation? Or perhaps we should put documents in-tree and use a script to collect them into a website or something.

**Comment by jdm** [comment]
I'm generally recommending that documentation go in the `docs` directory rather than the wiki, these days.

**Comment by shinglyu** [comment]
I see, perhaps a `COMMAND_LINE_ARGS.md` in which we can document the usage and effect of the arguments? This will be much more helpful than the one-liner from `./mach -h`.

**Comment by jdm** [comment]
That's not a bad idea. Documenting some of the most common and useful ones could be good, which would let us introduce --pref and direct the reader to prefs.json to learn more.

**Comment by shinglyu** [comment]
@danlrobertson I make a draft version, but I want to get your feedback before I submit a PR. I'm not sure if I'm getting the full size of the flex container in the correct way. 

https://github.com/servo/servo/compare/master...shinglyu:flex-reverse

Here is a screenshot, I don't know why the row-reverse one looks different if I run without `-o output.png`. 
![output](https://cloud.githubusercontent.com/assets/3250983/14878958/101c4220-0d59-11e6-8ae4-61d24c4d2f60.png)

**Comment by danlrobertson** [comment]
Great work! Column looks great, but there is a problem with row (unrelated to this issue). There shouldn't be padding between the blocks for `row-reverse`. Only real question I have is wouldn't the inline or block start be the accumulator - size of element? For example.

```rust
           let mut cur_b = self.block_flow.fragment.border_padding.block_end;
           for kid in &mut self.items {
               let base = flow::mut_base(flow_ref::deref_mut(&mut kid.flow));
               base.position.start.b = cur_b;
               cur_b = cur_b - base.position.size.block;
           }
```
Becomes
```rust
           let mut cur_b = self.block_flow.fragment.border_padding.block_end;
           for kid in &mut self.items {
               let base = flow::mut_base(flow_ref::deref_mut(&mut kid.flow));
               cur_b = cur_b - base.position.size.block;
               base.position.start.b = cur_b;
           }
```
I could be wrong.

**Comment by shinglyu** [comment]
Ah yes, your code looks more correct. I subtract the first kid's block size in the line 
```
base.position.start.b = cur_b - base.position.size.block;
```
, which is totally unnecessary if I swap the order of the two line. 

I notice the padding in `row-reverse`, I probably use the wrong full width, let me check that.

**Comment by shinglyu** [comment]
I tried fixing the `row-reverse` bug, now it looks better. But I'm not sure if I used the correct number, especially the [full width](https://github.com/servo/servo/compare/master...shinglyu:flex-reverse#diff-15e52e01e411afb4c1c5eed2c9b79706R274) and the [flex child's actual width](https://github.com/servo/servo/compare/master...shinglyu:flex-reverse#diff-15e52e01e411afb4c1c5eed2c9b79706R281)

![output](https://cloud.githubusercontent.com/assets/3250983/14909429/fb1f797e-0e15-11e6-960d-a8e259d869a3.png)

**Comment by shinglyu** [comment]
A stupid question, how to I properly run the css tests for flexbox?
I tried
`./mach test-css tests/wpt/css-tests/css-flexbox-1_dev` => Everything times out
`./mach test-css tests/wpt/css-tests/css-flexbox-1_dev/html/flexbox-flex-direction-column-reverse.htm` => Passes, but when I see with my naked eye I think it should fail
`./mach test-css tests/wpt/css-tests/css-flexbox-1_dev/html/reftest.list` => 0 test was ran

And do I need to use `--perf` when I run the tests?

**Comment by danlrobertson** [comment]
Good question! `./mach test-css tests/wpt/css-tests/css-flexbox-1_dev/html` is what I typically use. And you shouldn't need to run it with `--perf` options. See the `__dir__` file in the folder to check the `--pref` options the tests will be run with.

Looks great! I'd suggest submitting a PR soon so we can get more eyes on it.

**Comment by timvandermeij** [comment]
Fixed by #10987?
